packages/govuk-frontend/src/govuk/components/input/template.njk

```diff
@@ -6,6 +6,27 @@
   aria-describedby – for example hints or error messages -#}
 {% set describedBy = params.describedBy if params.describedBy else "" -%}
 
+{%- set hasPrefix = true if params.prefix and (params.prefix.text or params.prefix.html) else false %}
+{%- set hasSuffix = true if params.suffix and (params.suffix.text or params.suffix.html) else false %}
+
+{%- macro _inputElement(params) -%}
+  <input class="govuk-input {%- if params.classes %} {{ params.classes }}{% endif %} {%- if params.errorMessage %} govuk-input--error{% endif %}" id="{{ params.id }}" name="{{ params.name }}" type="{{ params.type | default("text", true) }}"
+    {%- if (params.spellcheck === false) or (params.spellcheck === true) %} spellcheck="{{ params.spellcheck }}"{% endif %}
+    {%- if params.value %} value="{{ params.value }}"{% endif %}
+    {%- if params.disabled %} disabled{% endif %}
+    {%- if describedBy %} aria-describedby="{{ describedBy }}"{% endif %}
+    {%- if params.autocomplete %} autocomplete="{{ params.autocomplete }}"{% endif %}
+    {%- if params.pattern %} pattern="{{ params.pattern }}"{% endif %}
+    {%- if params.inputmode %} inputmode="{{ params.inputmode }}"{% endif %}
+    {%- for attribute, value in params.attributes %} {{ attribute }}="{{ value }}"{% endfor %}>
+{%- endmacro -%}
+
+{%- macro _affixItem(affix, type) %}
+  <div class="govuk-input__{{ type }} {%- if affix.classes %} {{ affix.classes }}{% endif %}" aria-hidden="true" {%- for attribute, value in affix.attributes %} {{ attribute }}="{{ value }}"{% endfor %}>
+    {{- affix.html | safe | trim | indent(4) if affix.html else affix.text -}}
+  </div>
+{%- endmacro -%}
```

Comment:
Nunjucks is not where I'm the most familiar, are there any side effects to defining macros each time an input is rendered?

Reply:
Very little I imagine

It's the approach we started taking with Summary list, and now Button and Accordion:

https://github.com/alphagov/govuk-frontend/blob/8eeca8e8a7980dfaceb6619ca97084fdb63634f6/packages/govuk-frontend/src/govuk/components/summary-list/template.njk#L1

https://github.com/alphagov/govuk-frontend/blob/8eeca8e8a7980dfaceb6619ca97084fdb63634f6/packages/govuk-frontend/src/govuk/components/summary-list/template.njk#L13

https://github.com/alphagov/govuk-frontend/blob/8eeca8e8a7980dfaceb6619ca97084fdb63634f6/packages/govuk-frontend/src/govuk/components/button/template.njk#L23

https://github.com/alphagov/govuk-frontend/blob/8eeca8e8a7980dfaceb6619ca97084fdb63634f6/packages/govuk-frontend/src/govuk/components/accordion/template.njk#L3